src/events/upgrade.py

```diff
@@ -165,3 +170,26 @@ def _set_rolling_update_partition(self, partition: int) -> None:
             else:
                 cause = str(e)
             raise KubernetesClientError("Kubernetes StatefulSet patch failed", cause)
+
+    def apply_backwards_compatibility_fixes(self, event) -> None:
+        """A range of functions needed for backwards compatibility."""
+        logger.info("Applying upgrade fixes")
+        # Rev.38 (VM) - Create credentials for missing internal user, to reconcile state during upgrades
+        if (
+            not self.charm.state.cluster.internal_user_credentials
+            and self.charm.state.zookeeper.zookeeper_connected
+        ):
+            try:
+                internal_user_credentials = self.charm.zookeeper._create_internal_credentials()
+            except (KeyError, RuntimeError, subprocess.CalledProcessError, ExecError) as e:
+                logger.warning(str(e))
+                event.defer()
+                return
+
+            # only set to relation data when all set
+            for username, password in internal_user_credentials:
+                self.charm.state.cluster.update({f"{username}-password": password})
```

Comment:
**suggestion** which one will land first? I suggest you guys to decide which one you want to land first and coordinate in order to simplify your each other work. I'll definitely leave this up to you guys. I don't have any strong preference here, maybe just a small nudge: since this features would still require backporting to vm, in the process of backporting may anyway need to tweaking and changes (I reckon), whereas the CC features seems more aligned between VM and K8s, so I would prioritize that one. But again, up to you